### ISSUE TITLE
Not able to fetch the data by creating SimpleGraphQLRequest

### ISSUE THREAD

**Issue by sarvajeet294** [aws-amplify/amplify-android#1437]
I have created a query by using a key called "byActiveCustomers" where I am fetching the data in descending order of field "createdAt". I use following query to fetch the data on console which works perfectly.
query byActiveCustomers {
  customerByCreatedAt(sortDirection: DESC, is_active: 1, filter: {first_name: {beginsWith: "Sohail"}}) {
    items {
      id
      first_name
      last_name
      application_id
      mobile
      createdAt
      applications {
        branch_id
        amount
        agency_id
        agency_type_id
        agent_id
        application_number
        cdid
        createdAt
        id
        is_active
        is_assigned
        product_id
        remark
        status
        updatedAt
        user_id
      }
      cdid
      customer_type_id
      date_of_birth
      gender
      is_active
      phone
      updatedAt
    }
  }
}

Example Output of the above query when hit from console is as follows,
{
  "data": {
    "customerByCreatedAt": {
      "items": [
        {
          "id": "3794ff6f-0351-4338-9447-1c0a768b5c30",
          "first_name": "Sohail3",
          "last_name": "Ansari",
          "application_id": "b483f5b4-6a62-421a-9ecd-4507b1905784",
          "mobile": "6464646444",
          "createdAt": "2021-07-31T08:41:54.92Z",
          "applications": {
            "branch_id": "1",
            "amount": "569555",
            "agency_id": "1",
            "agency_type_id": "1",
            "agent_id": "1",
            "application_number": "",
            "cdid": "2e5f23b83a3207e104820210731141152",
            "createdAt": "2021-07-31T08:41:54.527Z",
            "id": "b483f5b4-6a62-421a-9ecd-4507b1905784",
            "is_active": 1,
            "is_assigned": 0,
            "product_id": "1a98f990-1286-42d8-b0ea-98a88546faa5",
            "remark": "",
            "status": "bfe01628-1a67-4826-899f-285c097d3abd",
            "updatedAt": "2021-07-31T08:49:21.115Z",
            "user_id": "1"
          },
          "cdid": "2e5f23b83a3207e104820210731141153",
          "customer_type_id": "3a6b41c8-a309-4026-99fe-fa5bcf395410",
          "date_of_birth": "03 Jul 2021",
          "gender": "3edae3d0-58d8-49f7-b86a-a1f8e37ccfdd",
          "is_active": 1,
          "phone": "53566226222",
          "updatedAt": "2021-07-31T08:49:21.387Z"
        }
      ]
    }
  }
}

But when I use this same query in my android app to create a SimpleGraphQLRequest by using following code,

String document = "query byActiveCustomers {\n" +
                "  customerByCreatedAt(sortDirection: DESC, is_active: 1, filter: {first_name: {beginsWith: \"Sohail\"}}) {\n" +
                "    items {\n" +
                "      id\n" +
                "      first_name\n" +
                "      last_name\n" +
                "      application_id\n" +
                "      mobile\n" +
                "      createdAt\n" +
                "      applications {\n" +
                "        branch_id\n" +
                "        amount\n" +
                "        agency_id\n" +
                "        agency_type_id\n" +
                "        agent_id\n" +
                "        application_number\n" +
                "        cdid\n" +
                "        createdAt\n" +
                "        id\n" +
                "        is_active\n" +
                "        is_assigned\n" +
                "        product_id\n" +
                "        remark\n" +
                "        status\n" +
                "        updatedAt\n" +
                "        user_id\n" +
                "      }\n" +
                "      cdid\n" +
                "      customer_type_id\n" +
                "      date_of_birth\n" +
                "      gender\n" +
                "      is_active\n" +
                "      phone\n" +
                "      updatedAt\n" +
                "    }\n" +
                "  }\n" +
                "}";

SimpleGraphQLRequest<List<Customer>> byActiveCustomer = new SimpleGraphQLRequest<>(document, Customer.class, new GsonVariablesSerializer());

        Amplify.API.query(byActiveCustomer,     //must call
                response -> {
                    List<Customer> paginated_application = response.getData();
                },
                error -> {
                    Log.e("MyAmplifyApp", "Query failure", error);
                }

I get a single entry of customer as response with response.getData() instead of list of customers. Also that customer contains all the fields as null.

response when hit from Android App:

GraphQLResponse{data='Customer {id=null, cdid=null, applications=null, customer_type_id=null, first_name=null, last_name=null, date_of_birth=null, gender=null, mobile=null, phone=null, is_active=null, createdAt=null, updatedAt=null}', errors='[]'}

**Comment by poojamat** [comment]
Try using the below syntax for GraphQLRequest parameter in the query method call:
```
 ModelQuery.list(Todo.class, Todo.NAME.contains("first")),
```
Something like:
```
    Amplify.API.query( ModelQuery.list(Customer.class, Customer.is_active.equals(1)),     //must call
            response -> {
                List<Customer> paginated_application = response.getData();
            },
            error -> {
                Log.e("MyAmplifyApp", "Query failure", error);
            }
```

**Comment by sarvajeet294** [comment]
Hi, the issue is not with the where condition. I have an issue where I need the data in sorted by 'createdAt' list. For that I have created the following key
@key(name: "byActiveCustomer", fields: ["is_active", "createdAt"], queryField: "customerByCreatedAt")

Which I am unable to to use anyhow? Please help me with finding a way to use this key which can give me the data in sorted list.

**Comment by raphkim** [comment]
It seems that the response from server is being serialized to an incorrect data type.
Please use the following snippet to generate your request object:
```
SimpleGraphQLRequest<PaginatedResult<Customer>> byActiveCustomer = new SimpleGraphQLRequest<>(
    document, 
    TypeMaker.getParameterizedType(PaginatedResult.class, Customer.class), 
    new GsonVariablesSerializer()
);
```

You may refer to the [following code sample](https://github.com/aws-amplify/amplify-android/issues/1438#issuecomment-892107605) for an example of how to correctly use `SimpleGraphQLRequest` to list models. Meanwhile, I encourage you to use `ModelQuery.list(...)` convenience method to generate a graphQL request object rather than using `SimpleGraphQLRequest` as @poojamat suggested to avoid confusion.

**Comment by sarvajeet294** [comment]
I tried the following code to fetch data in sorted by "createdAt" order,
SimpleGraphQLRequest<PaginatedResult<Customer>> byActiveCustomer = new SimpleGraphQLRequest<>(
    document, 
    TypeMaker.getParameterizedType(PaginatedResult.class, Customer.class), 
    new GsonVariablesSerializer()
);

But facing another issue as follows,
java.lang.NullPointerException: Attempt to invoke virtual method 'boolean com.google.gson.JsonElement.isJsonPrimitive()' on a null object reference
        at com.amplifyframework.api.aws.GsonGraphQLResponseFactory$IterableDeserializer.buildPaginatedResult(GsonGraphQLResponseFactory.java:159)

**Comment by abogatov** [comment]
Hi @raphkim 
> I encourage you to use ModelQuery.list(...) 
can you please explain how we can use the ModelQuery.list method to select data according to index (defined by @key in the GraphQL)? As @sarvajeet294 mentioned he created a key byActiveCustomer (this was supposed to create an index in dynamodb). How can we tell to the amplify generate a query with byActiveCustomer instead "list" statement?

**Comment by changxu0306** [comment]
Hi @abogatov ,
`ModelQuery.list` does the same thing with the query `byActiveCustomer`. It returns a GraphQLRequest which is built by `buildQuery` using predicate. Using `list` is a easier way to do the list query, so that you don't need to handle the serialization.
Here is the reference of `buildQuery` source code: 
[https://github.com/aws-amplify/amplify-android/blob/35f3b25a66154cc9dd83e4be15d52a50a8c69b1b/aws-api/src/main/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestFactory.java#L147](https://github.com/aws-amplify/amplify-android/blob/35f3b25a66154cc9dd83e4be15d52a50a8c69b1b/aws-api/src/main/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestFactory.java#L147)

**Comment by changxu0306** [comment]
> I tried the following code to fetch data in sorted by "createdAt" order,
> 
> ```
> SimpleGraphQLRequest<PaginatedResult<Customer>> byActiveCustomer = new SimpleGraphQLRequest<>(
>     document, 
>     TypeMaker.getParameterizedType(PaginatedResult.class, Customer.class), 
>     new GsonVariablesSerializer()
> );
> ```
> 
> But facing another issue as follows,
> 
> ```
> java.lang.NullPointerException: Attempt to invoke virtual method 'boolean com.google.gson.JsonElement.isJsonPrimitive()' on a null object reference
>         at com.amplifyframework.api.aws.GsonGraphQLResponseFactory$IterableDeserializer.buildPaginatedResult(GsonGraphQLResponseFactory.java:159)
> ```

Can you make sure you pass the correct format of document? An NPE appeared when calling `isJsonPrimitive()`method might due to your document didn't convert to Json successfully.

**Comment by abogatov** [comment]
> Hi @abogatov ,
> `ModelQuery.list` does the same thing with the query `byActiveCustomer`. It returns a GraphQLRequest which is built by `buildQuery` using predicate. Using `list` is a easier way to do the list query, so that you don't need to handle the serialization.
> Here is the reference of `buildQuery` source code:
> https://github.com/aws-amplify/amplify-android/blob/35f3b25a66154cc9dd83e4be15d52a50a8c69b1b/aws-api/src/main/java/com/amplifyframework/api/aws/AppSyncGraphQLRequestFactory.java#L147

Hi @changxu0306,
It doesn't seem to be the same thing.
I've analyzed the code and had a look at the AppSync request mapping template generated after pulling the GraphQL model. In case ModelQuery.list with predicate the amplify will use list<Model> query (not byActiveCustomer). According to the AppSync mapping template, the AppSync will do **scan data** in the dynamo DB and after it will apply the filter (predicate), so in this case, the index will not be used (the index is used in the mapping template for byActiveCustomer query where is used the query instead of scan). We will retrieve all data and filter them on the AppSync side instead of using the index of dynamo DB for retrieving need data (according to the index).
Is this a correct analysis?

In this regard, I have an additional question. If I use ownerField (in the @auth section) then the list method will retrieve all data from dynamoDB and filter it (apply auth rules) on the AppSync side for each record and every every time?

**Comment by raphkim** [comment]
Looks like we have a feature request already (#837) to cover your use-case of sorting query results by a filter. I will raise it once again with my team to find out current progress.

Meanwhile, it appears that our default Gson serializer isn't equipped to handle some response formats that deviate from our common-path use-cases. At this time, the only workaround I can suggest is to implement your own custom gson serializer to parse the GraphQL response into model objects.

**Comment by gpanshu** [comment]
Closing this issue as this will be resolved once #537 is done. Thank you. In the meantime I would recommend using the workaround that was suggested.
